edge-hub/watchdog/linux/src/main.rs

```diff
@@ -0,0 +1,183 @@
+use std::{
+    io::Error,
+    process::{exit, Child, Command, Stdio},
+    sync::atomic::{AtomicBool, Ordering},
+    sync::Arc,
+    thread,
+    time::Duration,
+};
+
+use nix::sys::signal::{self, Signal};
+use nix::unistd::Pid;
+use signal_hook::{flag, SIGINT, SIGTERM};
+use tracing::{error, info, subscriber, Level};
+use tracing_subscriber::fmt::Subscriber;
+
+const PROCESS_SHUTDOWN_TOLERANCE_SECS: u64 = 5;
+const PROCESS_POLL_INTERVAL_SECS: u64 = 1;
+
+struct ChildProcess {
+    pub name: String,
+    pub process: Child,
+}
+
+// TODO: find documentation standards for func param
+impl ChildProcess {
+    pub fn run_child_process(
+        name: String,
+        command: &mut Command,
+        has_parent_started_shutdown: Arc<AtomicBool>,
+    ) -> Arc<AtomicBool> {
+        // start child process before spawning new thread because it will get freed after the caller runs this function
+        // this makes lifetimes easier to deal with
+        let child = match command.spawn() {
+            Ok(child) => {
+                info!("Launched {:?} process with pid {:?}", name, child.id());
+                child
+            }
+            Err(e) => {
+                error!("Failed to start {:?} process. {:?}", name, e);
+                exit(2);
+            }
+        };
+
+        let mut has_shutdown = Arc::new(AtomicBool::new(false));
+        let has_shutdown_listener = Arc::clone(&mut has_shutdown);
+        thread::spawn(move || {
+            let mut child_process = ChildProcess {
+                name: name,
+                process: child,
+            };
+
+            while child_process.is_running() && !has_parent_started_shutdown.load(Ordering::Relaxed)
+            {
+                let poll_interval: Duration = Duration::new(PROCESS_POLL_INTERVAL_SECS, 0);
+                thread::sleep(poll_interval);
+            }
+
+            child_process.shutdown_if_running();
```

Comment:
I did not want to check the running/stopped status here and log because, even with this check here, we should still check before sending SIGTERM. So there is no reason to do the check twice.